src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.HighPriorityProcessor.cs

```diff
@@ -44,16 +44,17 @@ private sealed class HighPriorityProcessor : IdleProcessor
                         Start();
                     }
 
+                    public ImmutableArray<IIncrementalAnalyzer> Analyzers => _lazyAnalyzers.Value;
+
                     public Task Running => _running;
 
                     public int WorkItemCount => _workItemQueue.WorkItemCount;
                     public bool HasAnyWork => _workItemQueue.HasAnyWork;
 
                     public void AddAnalyzer(IIncrementalAnalyzer analyzer)
                     {
-                        var analyzers = _lazyAnalyzers.Value;
-
-                        _lazyAnalyzers = new Lazy<ImmutableArray<IIncrementalAnalyzer>>(() => analyzers.Add(analyzer));
+                        var analyzers = this.Analyzers;
+                        Interlocked.Exchange(ref _lazyAnalyzers, new Lazy<ImmutableArray<IIncrementalAnalyzer>>(() => analyzers.Add(analyzer)));
                     }
```

Comment:
this doesn't seem safe.  consider, two calls come through to AddAnalyzer at the same time.  both will capture a copy of this.Analyzers at the intial point.  Then both do an InterlockedExchange on the current field, but have a lazy that reference the original starting point.  so so whoever overwrites _lazyAnalyzers last will replace it with the original value plus themselves (thus missing the other guy).

Reply:
why are we ever using Interlocked like this instead of a simple lock.  Locks are really really great.  They are clear, prevent any sort of confusing overlapping, and are often extremely efficient.  They can occasionally be contention points in extremely hot-spots.  But unless AddAnalyzer is that sort of hotspot, this seems totaly like the wrong way to write ths.

Reply:
I am going to revert the change to this file. It seems there is another API already in place in the work coordinator to get all analyzers for high priority processor.

Reply:
Actually, my above comment https://github.com/dotnet/roslyn/pull/39699#discussion_r344395824 is not true. I am going to take your suggestion to move to using a lock.